modules/@apostrophecms/doc/index.js

```diff
@@ -1093,6 +1093,97 @@ module.exports = {
+      // This method also recursively invokes
+      // itself as needed for relationships nested in widgets,
+      // array fields and object fields.
+      //
+      // If a relationship field is present by name (such as `_products`)
+      // in the document, that is taken as authoritative, and any
+      // existing values in the `idsStorage` and `fieldsStorage`
+      // are overwritten. If the relationship field is not present, the
+      // existing values are left alone. This allows the developer
+      // to safely update a document that was fetched with
+      // `.relationships(false)`, provided the projection included
+      // the ids.
+      //
+      // If the `allRelatedDocsIds` argument is given, this function
+      // will only fill it with all the related docs ids, recursively.
+      //
+      // Currently `req` does not impact this, but that may change.
+
+      walkThrough(req, doc, allRelatedDocsIds) {
+        if (doc.metaType === 'doc') {
+          const manager = self.getManager(doc.type);
+          if (!manager) {
+            return;
+          }
+          forSchema(manager.schema, doc);
+        } else if (doc.metaType === 'widget') {
+          const manager = self.apos.area.getWidgetManager(doc.type);
+          if (!manager) {
+            return;
+          }
+          forSchema(manager.schema, doc);
+        }
+        function forSchema(schema, doc) {
+          for (const field of schema) {
+            if (field.type === 'area') {
+              if (doc[field.name] && doc[field.name].items) {
+                for (const widget of doc[field.name].items) {
+                  self.walkThrough(req, widget, allRelatedDocsIds);
+                }
+              }
+            } else if (field.type === 'array') {
+              if (doc[field.name]) {
+                doc[field.name].forEach(item => {
+                  item._id = item._id || self.apos.util.generateId();
+                  item.metaType = 'arrayItem';
+                  item.scopedArrayName = field.scopedArrayName;
+                  forSchema(field.schema, item);
+                });
+              }
+            } else if (field.type === 'object') {
+              const value = doc[field.name];
+              if (value) {
+                value.metaType = 'object';
+                value.scopedObjectName = field.scopedObjectName;
+                forSchema(field.schema, value);
+              }
+            } else if (field.type === 'relationship') {
+              if (Array.isArray(doc[field.name])) {
+                const relatedDocsIds = doc[field.name].map(relatedDoc => self.toAposDocId(relatedDoc));
+
+                if (Array.isArray(allRelatedDocsIds)) {
+                  allRelatedDocsIds.push(...relatedDocsIds);
+
+                  // Stop there, we only want to add the related IDs
+                  // to `allRelatedDocsIds` array reference, if provided, nothing more.
+                  return;
+                }
```

Comment:
That's really how I managed to reuse the code, without changing the whole function, since `doc` is mutated a lot in this recursion.

I'd be happy to have a better way to do that